### PR TITLE
wtdbg wrapper

### PR DESCRIPTION
**Pull request by mmiladi**

### CODE REVIEW

File: tools/wtdbg/test-data/dbg.1.dot [hunk 1]
```diff
@@ -0,0 +1,279 @@
+digraph {
+node [shape=record]
+N0 [label="{N0 42 | S0000000013 | R_8960_1024}"]
+N1 [label="{N1 41 | S0000000013 | R_7168_1024}"]
```
Comment (bgruening) [inline tools/wtdbg/test-data/dbg.1.dot]: please remove all of these files, I don't think they are necessary ...

File: tools/wtdbg/macros.xml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    <xml name="requirements">
+        <requirements>
+        <requirement type="package" version="1.2.8.1">wtdbg</requirement>
+            <yield/>
+        </requirements>
+    </xml>
+    <xml name="citations">
+        <citations>
+            <citation type="bibtex">
+@misc{githubwtdbg-1.2.8,
```
Comment (bgruening) [inline tools/wtdbg/macros.xml]: wouldn't it be better to cite the original repo?

File: tools/wtdbg/.shed.yml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+name: wtdbg
+owner: bgruening
+description: WTDBG is a fuzzy Bruijn graph (FBG) approach to long noisy reads assembly.
+homepage_url: https://github.com/ruanjue/wtdbg-1.2.8
```
Comment (bgruening) [inline tools/wtdbg/.shed.yml]: original repo?

File: tools/wtdbg/wtdbg.xml [hunk 4]
```diff
@@ -0,0 +1,195 @@
+<tool id="wtdbg" name="wtdbg" version="0.1.0">
```
Comment (bgruening) [inline tools/wtdbg/wtdbg.xml]: galaxy version same as tool version?

File: tools/wtdbg/wtdbg.xml [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    <macros>
+        <import>macros.xml</import>
+    </macros>
+    <expand macro="requirements" />
+    <command detect_errors="exit_code"><![CDATA[
+        wtdbg
+        -t \${GALAXY_SLOTS:-4} 
+        -i $i
```
Comment (bgruening) [inline tools/wtdbg/wtdbg.xml]: single quote all data and text params

File: tools/wtdbg/wtdbg.xml [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        <param argument="--tidy-reads" name="tidy_reads" type="integer" value="0" label="Filter reads less than tidy-reads" />
+        <param argument="--edge-min" name="edge_min" type="integer" value="3" label="The minimal depth of a valid edge set to" />
+        <param argument="--rescue-low-cov-edges" name="rescue_low_cov_edges" type="boolean" truevalue="--rescue-low-cov-edges" 
+          falsevalue="" label="Try to rescue low coverage edges" />
+
+        <section name="cns" title="Consensus options">
+            <!-- optional inputs -->
+            <!-- <param argument="-i" type="data" format="utg.cns" label="Input file(s) *.utg.cns" /> -->
+
+            <param argument="-j" type="integer" value="1000" label="Expected length of node" />
```
Comment (bgruening) [inline tools/wtdbg/wtdbg.xml]: can we add min/max to anyone of those?

File: tools/wtdbg/wtdbg.xml [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+            <param name="cns.E" value="-3"/>
+            <param name="cns.j" value="500"/>
+            <param name="cns.m" value="2"/>
+            <param name="cns.k" value="5"/>
+            <output name="output1" file="consensus_result3.fa"/>
+        </test>
+    </tests>
+    
+    <help><![CDATA[
+        WTDBG: De novo assembler for long noisy sequences
```
Comment (bgruening) [inline tools/wtdbg/wtdbg.xml]: this could be slightly improved :)

File: tools/wtdbg/wtdbg_kbm.xml [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    </outputs>
+    <tests>
+        <test>
+            <param name="input1" value="ecoli-ref.fa"/>
+            <param name="input2" value="ecoli-reads.fa"/>
+            <output name="output1" file="ecoli-reads_result1.kbmap"/>
+        </test>
+    </tests>
+    <help><![CDATA[
+        Program: kbm is a simple instance which implemented kmer-binmap
```
Comment (bgruening) [inline tools/wtdbg/wtdbg_kbm.xml]: improve help a little bit.

File: tools/wtdbg/wtdbg_kbm.xml [hunk 9]
```diff
@@ -0,0 +1,80 @@
+<tool id="wtdbg_kbm" name="wtdbg_kbm" version="0.1.0">
+    <macros>
+        <import>macros.xml</import>
+    </macros>
+    <expand macro="requirements" />
+    <command detect_errors="exit_code"><![CDATA[
+        kbm 
+        -t \${GALAXY_SLOTS:-4}
+        -d $input1
```
Comment (bgruening) [inline tools/wtdbg/wtdbg_kbm.xml]: single-quote

File: tools/wtdbg/wtdbg_kbm.xml [hunk 10]
```diff
@@ -0,0 +1,80 @@
+<tool id="wtdbg_kbm" name="wtdbg_kbm" version="0.1.0">
```
Comment (bgruening) [inline tools/wtdbg/wtdbg_kbm.xml]: description is missing

File: tools/wtdbg/wtdbg.xml (hunk 4 above)
Comment (bgruening) [inline tools/wtdbg/wtdbg.xml]: description is missing

Also we need to add to all out tool `version_command`

File: tools/wtdbg/macros.xml (hunk 2 above)
Comment (mmiladi) [reply tools/wtdbg/macros.xml]: This is the orig one! :)

File: tools/wtdbg/.shed.yml (hunk 3 above)
Comment (mmiladi) [reply tools/wtdbg/.shed.yml]: Yes it is :)

File: tools/wtdbg/wtdbg.xml (hunk 4 above)
Comment (mmiladi) [reply tools/wtdbg/wtdbg.xml]: what is it used for?

Comment (bgruening) [reply tools/wtdbg/wtdbg.xml]: https://docs.galaxyproject.org/en/latest/dev/schema.html#tool-version-command

Comment (bgruening) [reply tools/wtdbg/wtdbg.xml]: -help + grep?

File: tools/wtdbg/macros.xml (hunk 2 above)
Comment (bgruening) [reply tools/wtdbg/macros.xml]: I was referring to this one: https://github.com/ruanjue/wtdbg

File: tools/wtdbg/wtdbg.xml (hunk 6 above)
Comment (mmiladi) [reply tools/wtdbg/wtdbg.xml]: Those who had limits mentioned in help are already added.

File: tools/wtdbg/macros.xml (hunk 2 above)
Comment (mmiladi) [reply tools/wtdbg/macros.xml]: I know, but the other one is really "main" one. More help, recent updates, detailed description etc. Also the author has asked to refer githubwtdbg-1.2.8 in the readme.

File: tools/wtdbg/wtdbg.xml [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+            <param argument="-E" type="integer" value="-2" label="Gap extension score" />
+            <param argument="-m" type="integer" value="1" label="1: DBG correction; 2: DAG correction" />
+            <param argument="-S" type="integer" value="1" label="correct structure before error correction" />
+            <param argument="-v" type="boolean" truevalue="-v" falsevalue="" checked="False" label="Verbose" />
+        </section>
+
+    </inputs>
+    <outputs>
+        <data name="output_alignments" format="fasta" label="${tool.name}  alignments" from_work_dir="dbg.alignments" />
+        <data name="output_ctglay" format="text" label="${tool.name}  contigs layout" from_work_dir="dbg.ctg.lay" />
```
Comment (bgruening) [inline tools/wtdbg/wtdbg.xml]: txt is a valid format

File: tools/wtdbg/wtdbg.xml [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+            <param argument="-L" type="integer" value="10" label="High coverage cutoff" />
+            <param argument="-c" type="integer" value="0" label="Candidate strategy, 0: best-kmers, 1: median length, 2: first (include), 3: first (exclude), 4: longest, 5, shortest" />
+            <param argument="-M" type="integer" value="2" label="Match score" />
+            <param argument="-X" type="integer" value="-7" label="Mismatch score" />
+            <param argument="-I" type="integer" value="-3" label="Insertion score" />
+            <param argument="-D" type="integer" value="-4" label="Deletion score" />
+            <param argument="-E" type="integer" value="-2" label="Gap extension score" />
+            <param argument="-m" type="integer" value="1" label="1: DBG correction; 2: DAG correction" />
+            <param argument="-S" type="integer" value="1" label="correct structure before error correction" />
+            <param argument="-v" type="boolean" truevalue="-v" falsevalue="" checked="False" label="Verbose" />
```
Comment (bgruening) [inline tools/wtdbg/wtdbg.xml]: verbose can be removed

File: tools/wtdbg/wtdbg.xml [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+            <param argument="-W" type="integer" value="48" label="W-cutoff, drop the lagger (position)" />
+            <param argument="-H" type="integer" value="1" label="High coverage bonus" />
+            <param argument="-L" type="integer" value="10" label="High coverage cutoff" />
+            <param argument="-c" type="integer" value="0" label="Candidate strategy, 0: best-kmers, 1: median length, 2: first (include), 3: first (exclude), 4: longest, 5, shortest" />
+            <param argument="-M" type="integer" value="2" label="Match score" />
+            <param argument="-X" type="integer" value="-7" label="Mismatch score" />
+            <param argument="-I" type="integer" value="-3" label="Insertion score" />
+            <param argument="-D" type="integer" value="-4" label="Deletion score" />
+            <param argument="-E" type="integer" value="-2" label="Gap extension score" />
+            <param argument="-m" type="integer" value="1" label="1: DBG correction; 2: DAG correction" />
```
Comment (bgruening) [inline tools/wtdbg/wtdbg.xml]: This could be a select with two options.

File: tools/wtdbg/wtdbg.xml [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+            <!-- optional inputs -->
+            <!-- <param argument="-i" type="data" format="utg.cns" label="Input file(s) *.utg.cns" /> -->
+
+            <param argument="-j" type="integer" value="1000" label="Expected length of node" />
+            <param argument="-k" type="integer" value="15" label="Kmer size for long reads" />
+            <param argument="-Z" type="integer" value="4" label="Z-cutoff, drop the lower" />
+            <param argument="-W" type="integer" value="48" label="W-cutoff, drop the lagger (position)" />
+            <param argument="-H" type="integer" value="1" label="High coverage bonus" />
+            <param argument="-L" type="integer" value="10" label="High coverage cutoff" />
+            <param argument="-c" type="integer" value="0" label="Candidate strategy, 0: best-kmers, 1: median length, 2: first (include), 3: first (exclude), 4: longest, 5, shortest" />
```
Comment (bgruening) [inline tools/wtdbg/wtdbg.xml]: select field?

File: tools/wtdbg/wtdbg.xml [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+        <param argument="F" type="boolean" truevalue="-F" falsevalue="" checked="False" label="Filter low frequency kmers by a 4G-bytes array" />
+        <param argument="S" type="integer" value="4" label="Subsampling kmers, 1/S kmers are indexed" />
+        <param argument="X" type="integer" value="4" label="Max number of bin(256bp) in one gap" />
+        <param argument="Y" type="integer" value="4" label="Max number of bin(256bp) in one deviation" />
+        <param argument="x" type="integer" value="-7" label="penalty for BIN gap" />
+        <param argument="y" type="integer" value="-21" label="penalty for BIN deviation" />
+        <param argument="l" type="float" value="2048" label="Min length of alignment" />
+        <param argument="m" type="float" value="200" label="Min matched" />
+        <param argument="s" type="float" value="0.2" label="Max length variation of two aligned fragments" />
+        <param argument="v" type="boolean" truevalue="-v" falsevalue="" checked="False" label="Verbose" />
```
Comment (bgruening) [inline tools/wtdbg/wtdbg.xml]: verbose can be removed

File: tools/wtdbg/wtdbg_kbm.xml [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+        -d '$input1' 
+        -i '$input2' 
+        -o reads.kbmap
+    ]]></command>
+    <inputs>
+        <param type="data" name="input1" format="fasta" />
+        <param type="data" name="input2" format="fasta" />
+    </inputs>
+    <outputs>
+        <data name="output1" format="kbmap" from_work_dir="reads.kbmap" />
```
Comment (bgruening) [inline tools/wtdbg/wtdbg_kbm.xml]: I'm not sure what file format this is, but Galaxy does not know it.
You could use tabular if you like.

File: tools/wtdbg/wtdbg_kbm.xml (hunk 8 above)
Comment (mmiladi) [reply tools/wtdbg/wtdbg_kbm.xml]: Forgot this one sorry. wip

File: tools/wtdbg/.shed.yml [hunk 17]
```diff
@@ -0,0 +1,11 @@
...
+long_description: |
```
Comment (bgruening) [inline tools/wtdbg/.shed.yml]: If you have multiple tools, or planning to have them, make a `suite` out of it.

Comment (mmiladi) [reply tools/wtdbg/.shed.yml]: There won't be more tools. Would that be ok to keep it like this?

**Comment by mmiladi** [comment]
@bgruening Opps sorry, test1 can be uncommented. If u would like extra tests :)

**Comment by mmiladi** [comment]
@bgruening crashing is due to the conda binary. Nothing can be done here :) right?

**Comment by bgruening** [comment]
> @bgruening crashing is due to the conda binary. Nothing can be done here :) right?
This is also crashing?

**Comment by mmiladi** [comment]
How about this :roll_eyes:  ?

File: tools/wtdbg/wtdbg_kbm.xml (hunk 8 above)
Comment (mmiladi) [reply tools/wtdbg/wtdbg_kbm.xml]: Include please :)

**Comment by mmiladi** [comment]
@bgruening Parallel runs with kbm are in immature state. Decided to skip it for now at least.

**Comment by mmiladi** [comment]
@bgruening thanks a lot.

**Pull request merged**